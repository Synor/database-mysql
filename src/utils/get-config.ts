import { SynorError } from '@synor/core'
import { ConnectionString } from 'connection-string'
import { readFileSync } from 'fs'
import { resolve as resolvePath } from 'path'

type ConnectionConfig = import('mysql').ConnectionConfig
type TLSConnectionOptions = import('tls').ConnectionOptions

type SSLConfig = Pick<
  TLSConnectionOptions,
  'ca' | 'cert' | 'ciphers' | 'key' | 'passphrase' | 'rejectUnauthorized'
>

type SSLParams = Pick<
  SSLConfig,
  'ciphers' | 'passphrase' | 'rejectUnauthorized'
> & {
  ca?: string
  key?: string
  cert?: string
}

type MySQLDatabaseConfig = Required<Pick<ConnectionConfig, 'database'>> &
  Pick<
    ConnectionConfig,
    'host' | 'port' | 'user' | 'password' | 'multipleStatements' | 'ssl'
  >

type MySQLEngineConfig = {
  migrationRecordTable: string
}

export function getConfig(
  uri: string
): {
  databaseConfig: MySQLDatabaseConfig
  engineConfig: MySQLEngineConfig
} {
  try {
    const {
      protocol,
      hostname: host,
      port,
      user,
      password,
      path,
      params = {}
    } = new ConnectionString(uri, {
      params: {
        synor_migration_record_table: 'synor_migration_record'
      }
    })

    if (!protocol) {
      throw new Error(`[URI] missing: protocol!`)
    }

    if (protocol !== 'mysql') {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`)
    }

    const database = path && path[0]

    if (!database) {
      throw new Error('[URI] missing: database!')
    }

    let ssl: undefined | string | SSLConfig

    if (params.ssl) {
      try {
        const sslParams: SSLParams = JSON.parse(params.ssl)

        ssl = {
          ciphers: sslParams.ciphers,
          passphrase: sslParams.passphrase,
          rejectUnauthorized: sslParams.rejectUnauthorized
        }

        if (sslParams.ca) {
          ssl.ca = readFileSync(resolvePath(sslParams.ca))
        }
        if (sslParams.cert) {
          ssl.cert = readFileSync(resolvePath(sslParams.cert))
        }
        if (sslParams.key) {
          ssl.key = readFileSync(resolvePath(sslParams.key))
        }
      } catch (_) {
        ssl = params.ssl
      }
    }

    const databaseConfig: MySQLDatabaseConfig = {
      database,
      host,
      port,
      user,
      password,
      multipleStatements: true,
      ssl
    }

    const engineConfig: MySQLEngineConfig = {
      migrationRecordTable: params.synor_migration_record_table
    }

    return {
      databaseConfig,
      engineConfig
    }
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', error)
  }
}
