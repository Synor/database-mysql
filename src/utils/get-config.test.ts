import fs from 'fs'
import { getConfig } from './get-config'

describe('utils:getMySQLConfig', () => {
  let uri: Parameters<typeof getConfig>[0]

  beforeEach(() => {
    uri = 'mysql://root:root@127.0.0.1:3306/synor'
  })

  test('accepts mysql uri', () => {
    expect(getConfig(uri)).toMatchSnapshot()
  })

  test(`throws if uri protocol is not 'mysql:'`, () => {
    uri = uri.replace('mysql:', 'postgresql:')
    expect(() => getConfig(uri)).toThrow()
  })

  test('throws if protocol is missing', () => {
    uri = uri.replace('mysql:', '')
    expect(() => getConfig(uri)).toThrow()
  })

  test('throws if database is missing', () => {
    uri = uri.replace('/synor', '')
    expect(() => getConfig(uri)).toThrow()
  })

  test('throws if uri is malformed', () => {
    uri = 'mysql://@ _ @/synor'
    expect(() => getConfig(uri)).toThrow()
  })

  test('accepts custom migration record table name', () => {
    const tableName = 'migration_history'
    uri = `${uri}?synor_migration_record_table=${tableName}`
    expect(getConfig(uri).engineConfig.migrationRecordTable).toBe(tableName)
  })

  test('multipleStatements config is set', () => {
    expect(getConfig(uri).databaseConfig.multipleStatements).toBe(true)
  })

  describe('ssl config', () => {
    beforeEach(() => {
      uri = 'mysql://root:root@127.0.0.1:3306/synor'
    })

    test('accepts uri encoded string', () => {
      const ssl = 'Amazon RDS'
      uri = `${uri}?ssl=${encodeURIComponent(ssl)}`
      expect(getConfig(uri).databaseConfig.ssl).toBe(ssl)
    })

    test('accepts stringified json', () => {
      const ssl = { rejectUnauthorized: false }
      uri = `${uri}?ssl=${encodeURIComponent(JSON.stringify(ssl))}`
      expect(getConfig(uri).databaseConfig.ssl).toMatchObject(ssl)
    })

    test.each([
      ['ca', 'CA'],
      ['cert', 'CERT'],
      ['key', 'KEY']
    ])(`reads ssl.%s file content`, (key, content) => {
      jest
        .spyOn(fs, 'readFileSync')
        .mockImplementationOnce((v: any) => Buffer.from(`CONTENT:${v}`))

      const ssl = { [key]: content }
      uri = `${uri}?ssl=${encodeURIComponent(JSON.stringify(ssl))}`
      expect(
        ((getConfig(uri).databaseConfig.ssl as any)[key] as Buffer).toString()
      ).toMatch(new RegExp(`CONTENT:.+${content}`))
    })
  })
})
