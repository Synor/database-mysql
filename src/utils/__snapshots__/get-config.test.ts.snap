// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils:getMySQLConfig accepts mysql uri 1`] = `
Object {
  "databaseConfig": Object {
    "database": "synor",
    "host": "127.0.0.1",
    "multipleStatements": true,
    "password": "root",
    "port": 3306,
    "ssl": undefined,
    "user": "root",
  },
  "engineConfig": Object {
    "migrationRecordTable": "synor_migration_record",
  },
}
`;
